src/ray/raylet/node_manager.cc

```diff
@@ -235,20 +236,32 @@ void NodeManager::Heartbeat() {
   RAY_LOG(DEBUG) << "[Heartbeat] sending heartbeat.";
   auto &heartbeat_table = gcs_client_->heartbeat_table();
   auto heartbeat_data = std::make_shared<HeartbeatTableDataT>();
-  auto client_id = gcs_client_->client_table().GetLocalClientId();
-  const SchedulingResources &local_resources = cluster_resource_map_[client_id];
-  heartbeat_data->client_id = client_id.hex();
+  const auto &my_client_id = gcs_client_->client_table().GetLocalClientId();
+  SchedulingResources &local_resources = cluster_resource_map_[my_client_id];
+  heartbeat_data->client_id = my_client_id.binary();
   // TODO(atumanov): modify the heartbeat table protocol to use the ResourceSet directly.
   // TODO(atumanov): implement a ResourceSet const_iterator.
+  RAY_LOG(DEBUG) << "[Heartbeat] resources available: "
+                 << local_resources.GetAvailableResources().ToString();
   for (const auto &resource_pair :
-       local_resources.GetAvailableResources().GetResourceMap()) {
+      local_resources.GetAvailableResources().GetResourceMap()) {
     heartbeat_data->resources_available_label.push_back(resource_pair.first);
     heartbeat_data->resources_available_capacity.push_back(resource_pair.second);
   }
   for (const auto &resource_pair : local_resources.GetTotalResources().GetResourceMap()) {
     heartbeat_data->resources_total_label.push_back(resource_pair.first);
     heartbeat_data->resources_total_capacity.push_back(resource_pair.second);
   }
+  // TODO(atumanov): extract resource load information from the SchedulingQueues class.
+  local_resources.SetLoadResources(this->local_queues_.GetResourceLoad());
```

Comment:
If I understand this correctly, this line is `O(N)` where `N` is the number of queued tasks and the line is also repeated every `number of nodes / heartbeat interval` ms? Does it make sense to cache the result of `GetResourceLoad` inside `local_queues_`? I'm okay with deferring this to a future PR, but just wondering if you had thoughts about it.

Also, nit, but please remove the `this->` to stay consistent with style.

Reply:
Here I'm computing load once per heartbeat. Currently, load computation is O(n) in the number of *ready* tasks. I don't expect the ready queue size to be large. In fact, I expect it to be controlled by the overcommit factor (currently implicitly set to 1.0 in this PR). In such a case, the size of the ready queue is expected to be bound by (2f-1.0)*node.total. 
I did think about optimizing this by caching the load. This adds complexity, as we need to make sure we keep the load accurate  and update it on each task state transition. I think it will make more sense when (1) I refactor the task queue transitions to be internal to the class (as opposed to external), and (2) when we see this adding performance overhead (in its current form I don't expect it to, given a small bound on the ready queue size.)